src/api.py

```diff
@@ -68,21 +69,29 @@ def _complete_text(
             top_p=self.config.top_p,
             n=self.config.n_completions,
             echo=self.config.echo,
-            stop=self.config.stop,
+            stop=stopwords,
             presence_penalty=self.config.presence_penalty,
             frequency_penalty=self.config.frequency_penalty,
             best_of=self.config.best_of,
         )
         return [choice.text for choice in completion.choices]
 
+    def _flagged(self, responses: List[str]) -> bool:
+        input_text = "\n\n".join(responses)
+        moderation = openai.Moderation.create(input=input_text)
+        return moderation['results'][0]['flagged']
+
     def run(
         self, request: PluginRequest[BlockAndTagPluginInput]
     ) -> InvocableResponse[BlockAndTagPluginOutput]:
         """Run the text generator against all Blocks of text."""
 
         file = request.data.file
         for block in request.data.file.blocks:
-            generated_texts = self._complete_text(block.text)
+            user_id = self.context.user_id if self.context is not None else "testing"
+            generated_texts = self._complete_text(block.text, user=user_id)
+            if self.config.moderate_output and self._flagged(generated_texts):
```

Comment:
when we were discussing this, i had envisioned (perhaps incorrectly) pre-filtering based on the *prompts* content. this is error'ing on *generated* content, correct? do we also need to pre-filter "obscene" prompts ?

Reply:
Correct - this is filtering the output, not the input.  If we filter the input, we potentially de-risk being blocked by openai (unless people turn the filtering off), but we add another round-trip and model latency. I've also assumed that openAI just doesn't return stuff for offensive input like with DALLE, but I could be wrong.